### PR TITLE
feat: reintroduce automated changelog

### PR DESCRIPTION
**Pull request by unnawut**
Closes https://github.com/omgnetwork/releases/issues/14

## Overview

Automated changelog was dropped a few versions ago. Now that we're pushing for automated release, this PR reintroduces it back with a bit more style (categorizing the issues/PRs, plus the tagging process that should go with it).

## Changes

- [x] Configure `github-changelog-generator` with relevant filters and labels
- [x] Write down the steps required in the release process (see https://github.com/omgnetwork/devops/wiki/Elixir-omg-new-release-manual#prepare-changelog-and-version-file-changes)
- [x] Setup a blocking review bot to enforce proper PR label

### Q: Why tag the PRs rather than the issues?

Issues don't directly translate to code changes. Some issues are closed without further actions (clarification issues, duplicated issues, rejected requests, etc.), and some are issues that are addressed elsewhere. These closed issues, if show up on the changelog, can be mistaken as implemented changes.

Since the changelog should represent the code changes within the repo, tagging and collecting information from the merged PRs gives a more accurate view of the changes.

You can still tag issues for project management and other purposes, just that they won't be part of the changelog.

Examples of issues that don't necessarily relate to release or code changes: https://github.com/omgnetwork/elixir-omg/issues/1694, https://github.com/omgnetwork/elixir-omg/issues/1581, https://github.com/omgnetwork/elixir-omg/issues/1541, https://github.com/omgnetwork/elixir-omg/issues/1439, https://github.com/omgnetwork/elixir-omg/issues/1423.

## Testing

- Run `make changelog`. The generated readme is easier to view by going to the formatted view: 
image

- After this PR is merged, all PRs created after the merge should have the label enforced.

### CODE REVIEW

**Comment by unnawut** [comment]
Looks like only one tag will be picked. E.g. https://github.com/omgnetwork/elixir-omg/pull/1469 setting 3 changelog tags:
 
image

Only shows up in the "API changes" section and not "Breaking changes": https://github.com/omgnetwork/elixir-omg/blob/eb4be7d4056d6fc12f6fda86a195f7c53d0ed4f8/CHANGELOG.md#api-changes-2

**Review by boolafish — APPROVED** [review]
🔥

File: .github/workflows/enforce-changelog-labels.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      # When updating the labels here, also update the `configure-sections` of the `.github_changelog_generator` file
+      ONE_OF_LABELS: "api|enhancement|breaking|bug|chore|documentation"
+    steps:
+      - name: Check the PR for a changelog label
+        id: check-changelog-label
+        continue-on-error: true
+        run: |
+          set -o xtrace
+          # Using the issues API instead of pulls because it can return only the labels
+          curl "https://api.github.com/repos/omgnetwork/elixir-omg/issues/${{ github.event.pull_request.number }}/labels" \
```
Comment (boolafish) [inline .github/workflows/enforce-changelog-labels.yml:23]: minor: consider to use env var `GITHUB_REPOSITORY` to replace the `omgnetwork/elixir-omg` part to make the GH action script more portable.

**Review by boolafish — COMMENTED** [review]
one issue of this approval approach would be it decrease the human approval for 1. But we can increase the approval to 3 to cover this I guess.

**Comment by unnawut** [comment]
> one issue of this approval approach would be it decrease the human approval for 1. But we can increase the approval to 3 to cover this I guess.

For future reference, another approach I think of is making the bot dismisses its change request when passing. But it gets complicated because it needs to know the review_id which needs to be extracted and persisted across the builds. :(

So yeah, increasing the approval count after merge.

**Review by jarindr — APPROVED** [review]

File: .github/workflows/enforce-changelog-labels.yml (hunk 1 above)
Comment (unnawut) [reply .github/workflows/enforce-changelog-labels.yml]: https://github.com/omgnetwork/elixir-omg/pull/1708/commits/32fcd381b756529f5a52629b5f056482b30d7a9f

**Comment by unnawut** [comment]
> > one issue of this approval approach would be it decrease the human approval for 1. But we can increase the approval to 3 to cover this I guess.
> 
> For future reference, another approach I think of is making the bot dismisses its change request when passing. But it gets complicated because it needs to know the review_id which needs to be extracted and persisted across the builds. :(
> 
> So yeah, increasing the approval count after merge.

Actually we don't need the bot to request change or approve. It's a build step so I can just make it a required step. Updating...

**Comment by unnawut** [comment]
ok works now, apologies for all the noise

**Comment by unnawut** [comment]
Build step enforcement is enabled 
image

**Pull request merged**